### ISSUE TITLE
Save and Load performance

### ISSUE THREAD

**Issue by issackelly** [automerge/automerge#596]
I am investigating some thread exhaustion issues with my application and this has led me toward exploring some metrics in automerge.

I have a document (happy to share privately) of around 1MB in size, around 170k change operations. For purposes of this ticket, it's called `database.7.10.20230508`

In my application I call `document.save()` periodically, and on application exit to persist the document to disk in the event of crashes, backups, restarts, etc.

This call, with this database takes a (blocking) 0.7s [1], which seems like a pretty significant amount of time. Is it?

```
    pub fn whole_document_blob(&mut self) -> AutomergeSyncBlob {
        let start = Instant::now();
        let mut doc = self.document.write().unwrap();
        histogram!("wait_for_automerge_lock", start.elapsed());
        let start = Instant::now();
        let res = doc.save();
        histogram!("build_automerge_document_bytes", start.elapsed());
        res
    }
 ```

```
# TYPE build_automerge_document_bytes summary
build_automerge_document_bytes{quantile="0"} 0.706331986
build_automerge_document_bytes{quantile="0.5"} 0.706310733304708
build_automerge_document_bytes{quantile="0.9"} 0.706310733304708
build_automerge_document_bytes{quantile="0.95"} 0.706310733304708
build_automerge_document_bytes{quantile="0.99"} 0.706310733304708
build_automerge_document_bytes{quantile="0.999"} 0.706310733304708
build_automerge_document_bytes{quantile="1"} 0.72542781
build_automerge_document_bytes_sum 53.609582082
build_automerge_document_bytes_count 75
``` 

I think that I could possibly be using `save_incremental` for the periodic calls and only use `save` on load (so the first one comes through?) and on program exit but I'm not sure.

The docstring for this is not suggestive that this is correct (our data is more transactional and not like a single character change) 

> The output of this will not be a compressed document format, but a series of individual changes. This is useful if you know you have only made a small change since the last save and you want to immediately send it somewhere (e.g. you’ve inserted a single character in a text object).

On the other hand it suggests "the output will not be a compressed document format" and I don't know that we care about byte size of the document much at all. 

I'm not seeing examples of this in either automerge-repo or in this one, so I would like to file this ticket to figure out what the appropriate patterns might be
 
 
[1] An earlier version of this ticket had a typo suggesting "0.7ms" was the value that I was calling "a significant amount of time" -- the value is 700 milliseconds or 0.7 seconds

**Comment by issackelly** [comment]
I think that the `test_save_incremental` is suggesting the following expected behavior for a long-running document that is persisted to disk:

* A document should call `save` at least once and persist to file `foo`
* After that, calls to `save_incremental` can be appended directly to `foo`. Assuming the filesystem syncs then all is well.
* One might want to also call `save` when the program is exiting, to compress the changes, but it is not necessary.

**Comment by heckj** [comment]
Observer note - FWIW: 0.7ms (70 microseconds) seems like a pretty good response time to serializing out an entire data structure with the RLE encoding and such.

**Comment by issackelly** [comment]
Blerg. Thanksf for reading. I need to go double check all my units but I believe that was a typo and it's "750ms" or "0.75 seconds" not 75 micros.

**Comment by alexjg** [comment]
Some notes from investigating this:

1. You can shave a good 30% of the save time by using `Automerge::save_nocompress` (a terribly named method) which saves without DEFLATE compressing the columns. This does increase the file size quite a bit (doubles it in my experiments but it depends on your data)
2. I have a few small changes I'll put in a PR for which shave off another 20% or so by using a `HashMap` with `fxhash` instead of a `BTreeMap` for some data structures used when saving.

However, this is about as far as small tweaks will get us. The majority of the time is spent doing two things:

1. When encoding the change metadata (i.e. everything about a change except it's operations, which are encoded separately) for each change we encode the dependencies of the change as indices into the entire list of changes. We currently populate a `HashMap<ChangeHash, usize>` with the index of each change when we start saving and then for each change we look up the index of each of its dependencies in this map. For documents with large histories we spend around 50% of our time constructing this hashmap and then looking up dependencies from it.
2. Most of the rest of our time is spent allocating as we encode columns. This is because the backing memory for the columnar encoding infrastructure is just a vec, so every so often it has to allocate and copy as it grows. 

There are two directions which seem likely to speed these things up:

1. Find a way of storing the changes in an order which makes calculating the dependency indices less costly to calculate
2. Use some kind of bump allocator or similar for the encoding

**Comment by issackelly** [comment]
Thank you so much Alex. What about `save_incremental`? Is my description about how that should be used correct?

I'm curious if there are other approaches that might dramatically improve save and load (at the cost of cross-version incompatibilities) like.... persisting a big chunk of active memory to disk, maybe we _don't_ need to compute all of this stuff to save and load a document.

**Comment by alexjg** [comment]
Yep `save_incremental` is intended to be used for the majority of writes, appending to the original saved document. `save` could be called periodically to compress the log of changes.

Orion and I are cooking up a nicer API for incremental saving because right now the management of the "when did you last save" state is bound up with the state of the document when it would be useful to manage it separately.

But yes, ultimately we need ways of saving and loading a document which don't require the whole hash graph. There are a whole load of different approaches to this (some discussed tangentially in https://github.com/automerge/automerge/discussions/546#discussioncomment-5245812). There are also a number of other reasons we're going to want to redesign/extend the file format, such as branches, removing history, supporting gossip in the sync protocol etc. Basically, there's a bunch of fun new design work to do which I hope we can start soon.
